src/main/java/ch/njol/skript/util/BlockDataUtils.java

```diff
@@ -0,0 +1,330 @@
+
+import java.io.NotSerializableException;
+import java.io.StreamCorruptedException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.commons.lang.math.NumberUtils;
+import org.bukkit.Bukkit;
+import org.bukkit.Material;
+import org.bukkit.block.Block;
+import org.bukkit.block.BlockFace;
+import org.bukkit.block.data.BlockData;
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
+
+import ch.njol.skript.Skript;
+import ch.njol.util.StringUtils;
+import ch.njol.yggdrasil.Fields;
+import ch.njol.yggdrasil.YggdrasilSerializable.YggdrasilExtendedSerializable;
+
+/**
+ * Utility class for modifying {@link BlockData} of {@link Block Blocks}
+ */
+public class BlockDataUtils {
+	
+	private static final Map<String, StateType> stateTypes = new HashMap<>();
+	
+	static {
+		// Load all variations of block state types
+		if (Skript.isRunningMinecraft(1, 13)) {
+			for (Material material : Material.values()) {
+				// If this material is not a block, let's continue.
+				if (!material.isBlock())
+					continue;
+				
+				BlockData blockData = material.createBlockData();
+				String blockDataString = blockData.getAsString();
+				
+				// If this block data does not have states, let's continue.
+				if (!blockDataString.contains("["))
+					continue;
+				
+				String stateString = blockDataString.replaceAll(".+\\[|]", "");
+				String[] splits = stateString.split(",");
+				for (String s : splits) {
+					String[] split = s.split("=");
+					String state = split[0];
+					String name = state.replace("_", " ");
+					
+					if (!stateTypes.containsKey(name))
+						stateTypes.put(name, new StateType(name, state));
+					
+				}
+			}
+		}
+	}
+	
+	/**
+	 * Get all state types
+	 *
+	 * @return Collection of all state types
+	 */
+	public static Collection<StateType> getStateTypes() {
+		return stateTypes.values();
+	}
+	
+	/**
+	 * Get a string of all state type names
+	 * Used when registering class info
+	 *
+	 * @return String of all state type names
+	 */
+	public static String getStateTypeNames() {
+		return StringUtils.join(stateTypes.keySet(), ", ");
+	}
+	
+	/**
+	 * Get a state type by name
+	 *
+	 * @param name Name of state type
+	 * @return State type by name
+	 */
+	public static StateType getByName(String name) {
+		return stateTypes.get(name);
+	}
+	
+	/**
+	 * Change the state of a block
+	 *
+	 * @param block     Block to change
+	 * @param stateType StateType to change
+	 * @param value     New value
+	 */
+	public static void setBlockState(Block block, StateType stateType, Object value) {
+		BlockData blockData = block.getBlockData();
+		String key = block.getType().getKey().toString();
+		
+		try {
+			BlockData newData = Bukkit.createBlockData(key + "[" + stateType.type + "=" + value.toString().toLowerCase() + "]");
+			blockData = blockData.merge(newData);
+			block.setBlockData(blockData);
+		} catch (IllegalArgumentException ignore) {
```

Comment:
Nitpick, but I think this should just be ``} catch (IllegalArgumentException ignore) {}``, with the ending ``}`` on the same line as the catch.

Reply:
Yeah, again this is a draft. IntelliJ likes to move things around a lot.